src: add iterator for Object

RaisinTen opened the pull request:
Refs: https://github.com/nodejs/node-addon-api/issues/830

RaisinTen commented:
I made this very naive implementation of something that looks like an iterator for `Napi::Object`. Could we please discuss about how to implement this here? If I'm not wrong, in the C++ STL, RB trees are used to keep track of the sequence of iteration but I'm not sure what to use here other than what I have done below.

devsnek commented:
Can this work with c++ exceptions disabled?

RaisinTen commented:
> Can this work with c++ exceptions disabled?

Does all of `Napi::*` work when c++ exceptions are disabled?

devsnek commented:
yes, using NAPI_DISABLE_CPP_EXCEPTIONS

RaisinTen commented:
@devnek No error checking is implemented yet. I will keep in mind that it works without c++ exceptions.

gabrielschulhof commented:
@RaisinTen could you please add a test which preferably also includes a case where an exception is thrown during iteration?

RaisinTen commented:
@gabrielschulhof Do you mean a test case where the object `_object` is pointing to gets freed while iterating? Also, is there a way of getting a pointer to the underlying object keys array of the object that is being iterated? The current implementation keeps a copy of the object keys array when `std::begin()` is called on the object.

aminya commented:
> @gabrielschulhof Do you mean a test case where the object `_object` is pointing to gets freed while iterating? Also, is there a way of getting a pointer to the underlying object keys array of the object that is being iterated? The current implementation keeps a copy of the object keys array when `std::begin()` is called on the object.

I think a for ranged loop and maybe calling one of the STL functions like `std::sum` would be a good test case.

RaisinTen commented:
Since most of the tests check that the behaviour from the C++ side is coherent with the JS side, I wrote a test to check that the computed sum of a `Napi::Array` using the iterator is the same from both the JS side and the C++ side. Since I used the `for ranged loop` already, I didn't use `std::accumulate` on this.

Still not very sure, what I should do regarding exceptions though. Should we throw `std::out_of_range` when anything looks fishy inside the iterator?

RaisinTen commented on napi-inl.h:
```diff
@@ -1379,6 +1379,41 @@ inline void Object::AddFinalizer(Finalizer finalizeCallback,
+  return _index != other._index;
+}
+
+inline std::pair<Value, Value> Object::iterator::operator*() const {
+  const Value key = _keys.Get(_index);
+  const Value value = _object->Get(key);
+  return {key, value};
+}
```
This is technically still a `const_iterator`. I don't know how I could return an `std::pair` of references to the key and the value. Some help please?

RaisinTen commented on test/object/object.cc:
```diff
@@ -252,6 +252,17 @@ Value CreateObjectUsingMagic(const CallbackInfo& info) {
+  int64_t sum = 0;
+
+  for (const auto& e : array) {
+    sum += e.second.As<Number>().Int64Value();
```
Perhaps, I should change this test to iterate over an object because I'm planning on adding a separate iterator for `Napi::Array`, where the `Napi::Array::operator*()` would return the `Value` at the currently iterated index.

aminya replied:
Having a special iterator for arrays would be nice.

RaisinTen commented:
cc @gabrielschulhof, @mhdawson Could this please get another review?

legendecas commented on test/object/object.js:
```diff
@@ -157,4 +157,22 @@ function test(binding) {
   }
+
+  {
+    const obj = {
```
The test case here didn't reflect the exceptional path of JavaScript. Maybe we can test with proxies with throwing handlers:

```javascript
const o = new Proxy({}, {
  getOwnPropertyDescriptor(target, p) {
    // throwing primitives is valid in javascript;
    throw 'eh-oh';
  },
  ownKeys(target) {
    throw '¯\_(ツ)_/¯';
  },
});
```

RaisinTen replied:
@legendecas Running this locally produces this log:
```console
❯ node --trace-uncaught test/object/object.js

/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:175
        throw '¯\_(ツ)_/¯';
        ^
¯_(ツ)_/¯
Thrown at:
    at ownKeys (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:175:9)
    at test (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:180:23)
    at /home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:5:1
    at Module._compile (node:internal/modules/cjs/loader:1102:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1131:10)
    at Module.load (node:internal/modules/cjs/loader:967:32)
    at Module._load (node:internal/modules/cjs/loader:807:14)
    at executeUserEntryPoint (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
❯ echo $?
1
```
Should I write a test that asserts that this throws `/¯_\(ツ\)_\/¯/`?

legendecas replied:
The failure you encountered is probably the for-loop in JavaScript. What I mean is getting the proxied object to the native looping to see if the native side is handling the exception in a reasonable way.

RaisinTen replied:
@legendecas That is correct. I tried to call the function directly and it crashes because the iterator internally tries to get a copy of the keys of the object using `GetPropertyNames` and that throws the exception:

<details>

```console
❯ node test/object/object.js
FATAL ERROR: Error::Error napi_create_reference
 1: 0xacd720 node::Abort() [node]
 2: 0xaced2e node::OnFatalError(char const*, char const*) [node]
 3: 0xacedc1  [node]
 4: 0xa954f4 napi_fatal_error [node]
 5: 0xb5048821  [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 6: 0xb5048b0d Napi::Error::New(napi_env__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 7: 0xb508f5af GetPropertyNames(Napi::CallbackInfo const&) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 8: 0xb5096756 Napi::details::CallbackData<Napi::Array (*)(Napi::CallbackInfo const&), Napi::Array>::Wrapper(napi_env__*, napi_callback_info__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
...
```

</details>

However, this code too crashes and produces a similar trace if we add this in the same file (`test/object/object.js`):

```js
  {
    const o = new Proxy({}, {
      getOwnPropertyDescriptor(target, p) {
        // throwing primitives is valid in javascript;
        throw 'eh-oh';
      },
      ownKeys(target) {
        throw '¯\_(ツ)_/¯';
      },
    });
    binding.object.GetPropertyNames(o);
  }
```

<details>

```console
❯ node test/object/object.js
FATAL ERROR: Error::Error napi_create_reference
 1: 0xb37720 node::Abort() [node]
 2: 0xb38d2e node::OnFatalError(char const*, char const*) [node]
 3: 0xb38dc1  [node]
 4: 0xaff4f4 napi_fatal_error [node]
 5: 0xb5048821  [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 6: 0xb5048b0d Napi::Error::New(napi_env__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 7: 0xb508f5af GetPropertyNames(Napi::CallbackInfo const&) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 8: 0xb5096756 Napi::details::CallbackData<Napi::Array (*)(Napi::CallbackInfo const&), Napi::Array>::Wrapper(napi_env__*, napi_callback_info__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
...
```

</details>

Is this the expected behaviour?

legendecas replied:
Oh, my bad. Please note that it's the primitives that cause the crash. That is a known issue. The case you may need to take into consideration is _thwoing_ during the iteration:

```javascrpt
const o = new Proxy({}, {
  getOwnPropertyDescriptor(target, p) {
    throw new Error();
  },
  ownKeys(target) {
    throw new Error();
  },
});
```

RaisinTen replied:
@legendecas Thanks for updating the example. I tried to run this on `binding.object.sum()` and it produces the expected error:

<details>

```js
  {
    const obj = new Proxy({
      '-forbid': -0x4B1D,
      '-feedcode': -0xFEEDC0DE,
      '+office': +0x0FF1CE,
      '+forbid': +0x4B1D,
      '+deadbeef': +0xDEADBEEF,
      '+feedcode': +0xFEEDC0DE,
    }, {
      getOwnPropertyDescriptor(target, p) {
        throw new Error("getOwnPropertyDescriptor error");
      },
      ownKeys(target) {
        throw new Error("ownKeys error");
      },
    });

    binding.object.sum(obj);
  }
```

```console
❯ node test/object/object.js
/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:178
    binding.object.sum(obj);
                   ^

Error: ownKeys error
    at Object.ownKeys (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:174:15)
    at test (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:178:20)
    at Object.<anonymous> (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:5:1)
    at Module._compile (node:internal/modules/cjs/loader:1102:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1131:10)
    at Module.load (node:internal/modules/cjs/loader:967:32)
    at Function.Module._load (node:internal/modules/cjs/loader:807:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
```

</details>

But when I try to assert that it throws the particular error, the process seems to crash, which does not happen when I try to assert that `binding.object.GetPropertyNames(obj);` too throws the error.

<details>

```js
  {
    const obj = new Proxy({
      '-forbid': -0x4B1D,
      '-feedcode': -0xFEEDC0DE,
      '+office': +0x0FF1CE,
      '+forbid': +0x4B1D,
      '+deadbeef': +0xDEADBEEF,
      '+feedcode': +0xFEEDC0DE,
    }, {
      getOwnPropertyDescriptor(target, p) {
        throw new Error("getOwnPropertyDescriptor error");
      },
      ownKeys(target) {
        throw new Error("ownKeys error");
      },
    });

    assert.throws(() => {
      binding.object.sum(obj);
    }, /ownKeys error/);
  }
```

```console
❯ node test/object/object.js
FATAL ERROR: Error::New napi_get_last_error_info
 1: 0xb11720 node::Abort() [node]
 2: 0xb12d2e node::OnFatalError(char const*, char const*) [node]
 3: 0xb12dc1  [node]
 4: 0xad94f4 napi_fatal_error [node]
 5: 0xb4f5e841  [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 6: 0xb4f68a7a Napi::Error::New(napi_env__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 7: 0xb504589d Sum(Napi::CallbackInfo const&) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 8: 0xb4f68c70 Napi::details::CallbackData<Napi::Value (*)(Napi::CallbackInfo const&), Napi::Value>::Wrapper(napi_env__*, napi_callback_info__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
```

</details>

Since the current code does not attempt to catch any exception, shouldn't the exception propagate to the JS level anyway? I don't understand why there is a change in behaviour of the error when wrapping the thrower with an `assert.throws`.

legendecas replied:
The JavaScript exception is thrown at the `Object::iterator::iterator`. When c++ exceptions are disabled, the native code will continue to run if there are JavaScript exceptions thrown. And when JavaScript exception is pending, node-api would complain on most API calls, i.e. returning `napi_exceptions_pending`. So when `Object::iterator::iterator` triggered exceptions, the following steps like `Object::const_iterator::operator*` would abort subsequent `napi_exceptions_pending`.

RaisinTen replied:
@legendecas I tried to explicitly catch, clear and throw any pending exceptions in the dereferencing operator overloads but still I'm getting the same error:

<details>

```diff
diff --git a/napi-inl.h b/napi-inl.h
index c9f3ab1..54d7b52 100644
--- a/napi-inl.h
+++ b/napi-inl.h
@@ -1423,6 +1423,11 @@ inline std::pair<Value, Object::PropertyLValue<Value>>
 Object::const_iterator::operator*() const {
   Value key = _keys[_index];
   const PropertyLValue<Value> value = (*_object)[key];
+  Napi::Env env(_object->_env);
+  if (env.IsExceptionPending()) {
+    Error e = env.GetAndClearPendingException();
+    NAPI_THROW(e, {key, value});
+  }
   return {key, value};
 }
 
@@ -1459,6 +1464,11 @@ inline std::pair<Value, Object::PropertyLValue<Value>>
 Object::iterator::operator*() {
   Value key = _keys[_index];
   const PropertyLValue<Value> value = (*_object)[key];
+  Napi::Env env(_object->_env);
+  if (env.IsExceptionPending()) {
+    Error e = env.GetAndClearPendingException();
+    NAPI_THROW(e, {key, value});
+  }
   return {key, value};
 }
 
```

```console
❯ node test/object/object.js
FATAL ERROR: Error::New napi_get_last_error_info
 1: 0xb4b720 node::Abort() [node]
 2: 0xb4cd2e node::OnFatalError(char const*, char const*) [node]
 3: 0xb4cdc1  [node]
 4: 0xb134f4 napi_fatal_error [node]
 5: 0xb4e94871  [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 6: 0xb4e9eaaa Napi::Error::New(napi_env__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 7: 0xb4f7ba85 Sum(Napi::CallbackInfo const&) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 8: 0xb4e9eca0 Napi::details::CallbackData<Napi::Value (*)(Napi::CallbackInfo const&), Napi::Value>::Wrapper(napi_env__*, napi_callback_info__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
...
```

</details>

What am I doing wrong?

legendecas replied:
The key is that:

> When c++ exceptions are disabled, the native code will continue to run even if there are JavaScript exceptions thrown.

RaisinTen replied:
@legendecas But then, why does this run without crashing?

<details>

```js
  {
    const obj = new Proxy({
      '-forbid': -0x4B1D,
      '-feedcode': -0xFEEDC0DE,
      '+office': +0x0FF1CE,
      '+forbid': +0x4B1D,
      '+deadbeef': +0xDEADBEEF,
      '+feedcode': +0xFEEDC0DE,
    }, {
      getOwnPropertyDescriptor(target, p) {
        throw new Error("getOwnPropertyDescriptor error");
      },
      ownKeys(target) {
        throw new Error("ownKeys error");
      },
    });

    assert.throws(() => {
      binding.object.GetPropertyNames(obj);
    }, /ownKeys error/);
  }
```

</details>

`GetPropertyNames` too is being run with C++ exceptions disabled, so shouldn't the behaviour be the same?

legendecas replied:
The key is on "the native code will continue to run regardless of pending JavaScript exceptions". When there is only a single "continuous" node-api call, that's fine. When returning to node-core/v8 with pending JavaScript, the exception is routed to JavaScript. However, when there are multiple sequential continuous native node-api that calling into JavaScript and when the c++ exception is disabled, there is no way to break the native code execution on pending JavaScript, and those calls after the one caused the JavaScript exception would abort since they are unable to handle the situation - we can not throw again while there is a pending JavaScript exception.

Back to the case here, we have multiple continuous node-api calls in the iterators:
- one for getting object names (i.e., proxy handler `ownKeys`, https://github.com/nodejs/node-addon-api/pull/930/files#diff-cd0b229bcaf532206661e2ee1a603a7f60ac22ecfcb239920f6b42a34d909f34R1393) 
- and one for getting the first value of the keys array (https://github.com/nodejs/node-addon-api/pull/930/files#diff-cd0b229bcaf532206661e2ee1a603a7f60ac22ecfcb239920f6b42a34d909f34R1424)

Those two procedures are both located in node-addon-api so there is no way for user code to break on exceptions. There comes the crash.

RaisinTen replied:
@legendecas Thank you for the awesome explanation. I will enclose the iterator code in `#ifdef NAPI_CPP_EXCEPTIONS` guards. How do I locally build and test this with C++ exceptions enabled?

legendecas replied:
By default `npm test` will run the tests on both flavors. See https://github.com/nodejs/node-addon-api/blob/main/test/binding.gyp for build configurations and https://github.com/nodejs/node-addon-api/blob/main/test/common/index.js#L78 for test runners.

RaisinTen replied:
@legendecas Hi! I have made the changes so that the iterator is available only when we are building with c++ exceptions enabled and it passes the tests locally. PTAL.

RaisinTen replied:
ehhh... I was using the older test configuration while testing locally, that's why it is failing to require the module on GH Actions. Any neat way to tell the JS test whether the c++ exceptions is enabled?

legendecas replied:
> Any neat way to tell the JS test whether the c++ exceptions is enabled?

In `test/common/index.js` there is a function named `runTest` that enumerated all available build flavors to run tests. You can create a subset or options to filter out the c++ exceptions enabled flavors to run the new cases.

RaisinTen replied:
@legendecas Thank you. I've updated my change. CI is green now.

RaisinTen commented on napi.h:
```diff
@@ -598,6 +598,14 @@ namespace Napi {
     );
 
+    /// Gets or sets an indexed property or array element.
+    PropertyLValue<Value> operator[](Value index  /// Property / element index
```
@gabrielschulhof I agree, it does look strange. It was done by the linter though. I used `///` for the comments to make it consistent with the rest of the comments nearby. Perhaps we can change all comments to use `//` in a separate PR?

RaisinTen commented on napi.h:
```diff
@@ -772,6 +780,21 @@ namespace Napi {
+
+    inline iterator end();
+#endif  // NAPI_CPP_EXCEPTIONS
+
```
@gabrielschulhof The iterator classes require the definition of the `Napi::Array` class to be completed but it is not completed here yet. That's why I decided to move the actual class definitions after the definition of `Napi::Array`.

KevinEady commented on test/common/index.js:
```diff
@@ -78,15 +78,16 @@ exports.mustNotCall = function(msg) {
+    require.resolve(`../build/${buildType}/binding.node`);
+  const binding_noexcept =
+    require.resolve(`../build/${buildType}/binding_noexcept.node`);
 
-  for (const item of bindings) {
-    await Promise.resolve(test(require(item)))
-      .finally(exports.mustCall());
-  }
+  await Promise.resolve(test(require(binding), true))
+    .finally(exports.mustCall());
+
+  await Promise.resolve(test(require(binding_noexcept), false))
+    .finally(exports.mustCall());
```
This here

RaisinTen replied:
Done

KevinEady commented on test/object/object.js:
```diff
@@ -4,7 +4,7 @@ const assert = require('assert');
 module.exports = require('../common').runTest(test);
 
-function test(binding) {
+function test(binding, NAPI_CPP_EXCEPTIONS) {
```
And this

RaisinTen replied:
Done

KevinEady commented on test/object/object.js:
```diff
@@ -156,4 +156,48 @@ function test(binding) {
   }
+
+  {
+    if (NAPI_CPP_EXCEPTIONS) {
```
Maybe you can see if `'sum' in binding.object`?

KevinEady replied:
Also, this `if` statement seems to be inside an unnecessary additional block?

RaisinTen replied:
@KevinEady Thank you for the comments, addressed! :)

KevinEady reviewed (commented):
I'm not a big fan of the changes made in order to see if the binding has C++ exceptions enabled. Is it possible to change the logic?

RaisinTen commented:
Been a while since the last review. Can this please get another review as I have addressed all the comments and rebased?
cc @gabrielschulhof @KevinEady @legendecas @mhdawson

mhdawson commented:
@RaisinTen I looked through it and the main thing I think it needs is additions to the documentation. @gabrielschulhof @KevinEady @legendecas  anything else you want to point out?

RaisinTen commented:
@mhdawson added docs, PTAL :)

mhdawson commented on doc/object.md:
```diff
@@ -288,5 +288,47 @@ Napi::Value Napi::Object::operator[] (uint32_t index) const;
+Iterators expose an `std::pair<...>`, where the `first` property is a
+[`Napi::Value`](value.md) that holds the currently iterated key and the
+`second` property is a [`Napi::Object::PropertyLValue`](propertylvalue.md) that
+holds the currently iterated value. Iterators are available if C++ exceptions
```
```suggestion
holds the currently iterated value. Iterators are only available if C++ exceptions
```

RaisinTen replied:
Thanks, updated!

mhdawson reviewed (approved):
LGTM with a suggestion

devnek commented:
> @devnek No error checking is implemented yet. I will keep in mind that it works without c++ exceptions.

Please don't tag @devnek

mhdawson commented:
@NickNaso I think you wanted to do a pass through before this landed.

NickNaso commented:
> @NickNaso I think you wanted to do a pass through before this landed.

Yes, I'm taking a look right now.

NickNaso commented on doc/object.md:
```diff
@@ -288,5 +288,47 @@ Napi::Value Napi::Object::operator[] (uint32_t index) const;
+### Constant Iterator
+
+In constant iterators, the iterated values are immutable.
+
```
Here you should document the interface exposed by `Napi::Object::const_iterator`.

NickNaso commented on doc/object.md:
````diff
@@ -288,5 +288,47 @@ Napi::Value Napi::Object::operator[] (uint32_t index) const;
+
+In constant iterators, the iterated values are immutable.
+
+```cpp
````
Here you should add title `Example`

NickNaso commented on doc/object.md:
```diff
@@ -288,5 +288,47 @@ Napi::Value Napi::Object::operator[] (uint32_t index) const;
+### Non Constant Iterator
+
+In non constant iterators, the iterated values are mutable.
+
```
Here you should document the interface exposed by `Napi::Object::terator`

NickNaso commented on doc/object.md:
````diff
@@ -288,5 +288,47 @@ Napi::Value Napi::Object::operator[] (uint32_t index) const;
+
+In non constant iterators, the iterated values are mutable.
+
+```cpp
````
Here you should add title `Example`.

NickNaso reviewed (commented):
@RaisinTen just some notes about documentation.

RaisinTen commented:
@NickNaso thank you very much for the review comments. Before I address the comments, I would like to know whether I should add the changes in a separate PR because the changes in this PR seems to be landed already in https://github.com/nodejs/node-addon-api/commit/bd8f6e6d1a2259eeafa292a15e3991ad3587d9b1.
cc @mhdawson for some clarification.

mhdawson commented:
@RaisinTen thanks for noticing that. I think I landed it accidentally as part of landing another PR.  I think  making those changes in a new PR is the best way to go.
